### ISSUE TITLE
[Idea] Improve installing apps from web browser by using URL `flatpak://`

### ISSUE THREAD

**Issue by pktiuk** [flathub/website#919]
Currently `Install` button on flathub website is used for downloading `.flatpakref` file, then we have to open that file and install it on our system.

![obraz](https://user-images.githubusercontent.com/45544416/211215527-be3c29bd-bf2d-4120-aa5b-2644ab205660.png)

I think the installation process could be more streamlined if it used a URL with the specified protocol, for example `flatpak://`, which would open appstore used by the current system (like `gnome-software` or `discovery`).

**Comment by razzeee** [comment]
Would need implementation in software and/or discovery first. Also the url would need to embed the remote.

**Comment by pktiuk** [comment]
Yes, it would need, but in terms of implementation, this will be rather easy. URL should contain a similar set of data to `.flatpakref`, and local software stores could easily use these data to show proper app entry.

**Comment by JakobDev** [comment]
I'm also a fan of this. This would make the download button more useful. I think the best Idea would be, to just provide the link to the flatpakref e.g. `flatpak:RefUrl/https%3A//dl.flathub.org/repo/appstream/org.mozilla.firefox.flatpakref` or `flatpakref:https%3A//dl.flathub.org/repo/appstream/org.mozilla.firefox.flatpakref`. With this, the Software Centers just need to download the Ref and could use existing Code after that, so this should be easy to implement.

> Would need implementation in software and/or discovery first.

That's true. For KDE, I can think I should ping @aleixpol, for Gnome, I don't know who to ping.

**Comment by razzeee** [comment]
@pwithnall maybe

**Comment by pwithnall** [comment]
I thought the original intention with these files was to have your browser open them in gnome-software/Discovery automatically on download, which would give the desired behaviour?

I don’t know if that’s the default any more (I’ve messed my system around too much to know what’s default and what’s me), but that sounds a lot easier than standardising a new `flatpak:` URI scheme, implementing support for it in gnome-software and Discover, and making sure they both securely and correctly download the `.flatpakref` file they’re pointed to.

**Comment by JakobDev** [comment]
> I thought the original intention with these files was to have your browser open them in gnome-software/Discovery automatically on download, which would give the desired behaviour?

This is the intention, but on most Systems the file is just downloaded and opened by the Browser. The User have to open the file self.

**Comment by pwithnall** [comment]
Well could we configure the browser to not do that?

**Comment by JakobDev** [comment]
> Well could we configure the browser to not do that?

That would mean, that we need to configure dozens of different Browser with different install sources. That is unrealistic. Adding a URL Schema would make way more sense. And it is easy to impelemnt. Here is a example implementation in Bash:
```bash
#!/bin/bash

function urldecode() { : "${*//+/ }"; echo -e "${_//%/\\x}"; }

if  [[ $1 != flatpakref:* ]] ;
then
    exit 1
fi

url=$(urldecode "${1:11}")

tempfile=$(mktemp --suffix .flatpakref)

curl $url > $tempfile

xdg-open $tempfile
```
I'm sure this also can be implemented easy in Discover and Gnome Software and that would save a lot of trouble.

**Comment by razzeee** [comment]
For context, this worked, but stopped working at some point, at least with Firefox

**Comment by bragefuglseth** [comment]
No need for a new protocol, [Appstream URIs](https://www.freedesktop.org/software/appstream/docs/sect-AppStream-Misc-URIHandler.html) do exactly this

**Comment by bragefuglseth** [comment]
Oh, I see that it has already been mentioned. My bad!

**Comment by fcrozat** [comment]
An easier and faster way to handle this would be to configure Flathub http server to serve flatpakref file with the proper mimetype `application/vnd.flatpak.ref`. Currently, nginx is sending application/octet-stream.
This would ensure GNOME Software (on GNOME system) would be proposed by the web browser when people try to download flatpak ref files.

**Comment by razzeee** [comment]
> An easier and faster way to handle this would be to configure Flathub http server to serve flatpakref file with the proper mimetype `application/vnd.flatpak.ref`. Currently, nginx is sending application/octet-stream. This would ensure GNOME Software (on GNOME system) would be proposed by the web browser when people try to download flatpak ref files.

Just tried that, doesn't seem to change a thing

**Comment by pwithnall** [comment]
Draft implementation available for gnome-software here: https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/2089

**Comment by razzeee** [comment]
It's tracked for elementary here https://github.com/elementary/sideload/issues/169

**Comment by regs01** [comment]
> No need for a new protocol, [Appstream URIs](https://www.freedesktop.org/software/appstream/docs/sect-AppStream-Misc-URIHandler.html) do exactly this

Protocol is much easier and more straight forward to work with.

**Comment by razzeee** [comment]
The proposed is now supported in kde discover, gnome software and elementary OS. So we'll need for some releases and adoption of those, but hopefully it won't be too long.

**Comment by regs01** [comment]
What is live isn't exactly what is this issue is about. Its would be much simpler for users, if it was a simple protocol with simple scheme. And install and download buttons (or install button with dropdown with download option) on flathub to retain compatibility for a while.


> Would need implementation in software and/or discovery first. Also the url would need to embed the remote.

Clients always come last. In order to support something, there should be something.

**Comment by bbb651** [comment]
> What is live isn't exactly what is this issue is about. Its would be much simpler for users, if it was a simple protocol with simple scheme.

It's a simple protocol with a simple scheme, it's just https which points to a `.flatpakref` file which makes it incredibly simple to support for clients (it's not feasible to fit everything in the URI itself without an indirection without dropping the GPG key).

> And install and download buttons (or install button with dropdown with download option) on flathub to retain compatibility for a while.

They do in the page you get redirected to, although it would be nice to have in the dropdown as well.

> Clients always come last. In order to support something, there should be something.

Here I think it makes more sense to have support from clients first because like you've said it can be a bit disruptive if your client doesn't support it and due to new versions of clients usually taking a while to make it to users being distro packages.

**Comment by razzeee** [comment]
Kde, gnome and elementary have been shipping support for a while now. Not sure if those reached lts versions, but not willing to wait years.
